fw/tcp.c

```diff
@@ -50,59 +50,69 @@ tfw_tcp_propagate_dseq(struct sock *sk, struct sk_buff *skb)
 		tcb_next->end_seq++;
 }
 
+/*
+ * Setup all necessary fields for a new skb and insert it in the
+ * sk write queue. There is a case when we allocate new skb during
+ * processing skb in `xmit` callback. We should setup this new skb
+ * and insert it into skb write queue in right way. Most of code
+ * below was taken from `tso_fragment/tcp_fragment` functions,
+ * besides using `tfw_tcp_propagate_dseq`.
+ */
 void
 tfw_tcp_setup_new_skb(struct sock *sk, struct sk_buff *skb,
-		      unsigned int mss_now, bool tcp_fragment)
+		      struct sk_buff *nskb, unsigned int mss_now,
+		      bool tcp_fragment)
```

Comment:
Why can't we check `skb->len != skb->data_len` inside the function?

Reply:
Yes we can, fixed